### PR TITLE
Add implementation for 'Show message request' LS feature

### PR DESCRIPTION
**Pull request by evidolob**
### What does this PR do?
Add implementation for 'Show message request' LS feature

### What issues does this PR fix or reference?
#1794

#### Changelog
Add implementation for [ShowMessage Request](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#showmessage-request) Language Server protocol feature

#### Release Notes
Add implementation for [ShowMessage Request](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#showmessage-request) Language Server protocol feature

### CODE REVIEW

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/window/dialog/MessageDialogFooter.ui.xml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        .action-button {
+            margin-right: 1em;
+        }
+    </ui:style>
+
+    <g:HTMLPanel addStyleNames="{style.button-block}">
+    </g:HTMLPanel>
+</ui:UiBinder>
```
Comment (vparfonov) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/window/dialog/MessageDialogFooter.ui.xml]: new line

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/window/dialog/MessageDialogPresenter.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+
+    @Inject
+    public MessageDialogPresenter(final @NotNull MessageDialogView view) {
+        this.view = view;
+
+        this.view.setDelegate(this);
+    }
+
+//    @Override
```
Comment (vparfonov) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/window/dialog/MessageDialogPresenter.java]: Do we need this comment code fragment?

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/window/dialog/MessageDialogPresenter.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    public void onAction(MessageActionItem actionItem) {
+        view.closeDialog();
+        callback.accept(actionItem);
+    }
+
+    @Override
+    public void onEnterClicked() {
+//        if (view.isOkButtonInFocus()) {
```
Comment (vparfonov) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/window/dialog/MessageDialogPresenter.java]: Do we need this comment code fragment?

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/window/dialog/MessageDialogViewImpl.ui.xml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            flex-flow: column;
+            -webkit-align-items: stretch;
+            align-items: stretch;
+        }
+    </ui:style>
+
+    <g:HTMLPanel addStyleNames="{style.content-block}" debugId="dialog-confirm">
+        <g:SimplePanel ui:field="content" addStyleNames="{style.content}"/>
+    </g:HTMLPanel>
+</ui:UiBinder>
```
Comment (vparfonov) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/window/dialog/MessageDialogViewImpl.ui.xml]: new line

**Comment by tolusha** [comment]
Add `@author` annotation where it is absent.

**Comment by tsmaeder** [comment]
@evidolob just wondering where the test launguage server is coming from?

**Comment by benoitf** [comment]
@tsmaeder AFAIK https://github.com/eclipse/che-lib/pull/33

File: agents/pom.xml [hunk 5]
```diff
@@ -38,4 +38,12 @@
         <module>ls-typescript</module>
         <module>ls-csharp</module>
     </modules>
+    <profiles>
+        <profile>
+            <id>test-ls</id>
+            <modules>
```
Comment (skabashnyuk) [inline agents/pom.xml]: what are the purpose of this profile?

File: agents/test-ls/src/main/java/org/eclipse/che/api/agent/TestLSAgent.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import com.google.inject.Singleton;
+
+import org.eclipse.che.api.agent.shared.model.Agent;
+import org.eclipse.che.api.agent.shared.model.impl.BasicAgent;
+
+import java.io.IOException;
+
+/**
+ * Test Language server agent.
+ *
```
Comment (skabashnyuk) [inline agents/test-ls/src/main/java/org/eclipse/che/api/agent/TestLSAgent.java]: too many spaces.

File: agents/test-ls/src/main/java/org/eclipse/che/api/agent/TestLSModule.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+package org.eclipse.che.api.agent;
+
+import com.google.inject.AbstractModule;
+import com.google.inject.multibindings.Multibinder;
+
+import org.eclipse.che.api.agent.shared.model.Agent;
+import org.eclipse.che.inject.DynaModule;
+
+/**
+ *
```
Comment (skabashnyuk) [inline agents/test-ls/src/main/java/org/eclipse/che/api/agent/TestLSModule.java]: ?

File: core/che-core-api-core/src/main/java/org/eclipse/che/api/core/jsonrpc/impl/JsonRpcModule.java [hunk 8]
```diff
@@ -46,4 +49,10 @@ protected void configure() {
     public JsonParser jsonParser() {
         return new JsonParser();
     }
+
+    @Provides
+    @Singleton
+    protected Gson gson() {
+        return DtoFactory.getInstance().getGson();
```
Comment (skabashnyuk) [inline core/che-core-api-core/src/main/java/org/eclipse/che/api/core/jsonrpc/impl/JsonRpcModule.java]: very strange way to bind gson

File: plugins/plugin-languageserver/che-plugin-test-ls/src/main/java/org/eclipse/che/plugin/test/ls/inject/TestLSModule.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.eclipse.che.api.languageserver.launcher.LanguageServerLauncher;
+import org.eclipse.che.api.languageserver.shared.model.LanguageDescription;
+import org.eclipse.che.inject.DynaModule;
+import org.eclipse.che.plugin.test.ls.languageserver.TestLanguageServerLauncher;
+
+import static java.util.Arrays.asList;
+
+/**
+ *
```
Comment (skabashnyuk) [inline plugins/plugin-languageserver/che-plugin-test-ls/src/main/java/org/eclipse/che/plugin/test/ls/inject/TestLSModule.java]: ?

File: wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/messager/ShowMessageJsonRpcTransmitter.java [hunk 10]
```diff
@@ -55,4 +67,49 @@ private void configureUnSubscribeHandler(RequestHandlerConfigurator requestHandl
...
+                      .withConsumer(showMessageRequestEndpointIds::remove);
+    }
+
+    public CompletableFuture<MessageActionItem> sendShowMessageRequest(ShowMessageRequestParams requestParams) {
+        CompletableFuture<MessageActionItem> result = new CompletableFuture<>();
+        if (showMessageRequestEndpointIds.isEmpty()) {
+            result.complete(null);
+        }
+//        if (showMessageRequestEndpointIds.size() > 1) {
+//            result.completeExceptionally(new Exception("Can't send show message request, too meany clients"));
```
Comment (skabashnyuk) [inline wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/messager/ShowMessageJsonRpcTransmitter.java]: what is it?

File: wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/registry/CheLanguageClient.java [hunk 11]
```diff
@@ -62,18 +68,18 @@ public void publishDiagnostics(PublishDiagnosticsParams diagnostics) {
...
+                LOG.error(message.getMessage());
+                break;
+            case Warning:
+                LOG.warn(message.getMessage());
+                break;
+            case Info:
+                LOG.info(message.getMessage());
+                break;
+            case Log:
+                LOG.debug(message.getMessage());
```
Comment (skabashnyuk) [inline wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/registry/CheLanguageClient.java]: default?

File: wsagent/che-core-api-languageserver/src/test/java/org/eclipse/che/api/languageserver/LocalTestLSLauncher.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+import org.eclipse.che.api.languageserver.registry.LanguageServerDescription;
+import org.eclipse.lsp4j.jsonrpc.Launcher;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4j.services.LanguageServer;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ *
```
Comment (skabashnyuk) [inline wsagent/che-core-api-languageserver/src/test/java/org/eclipse/che/api/languageserver/LocalTestLSLauncher.java]: ?

File: wsagent/che-core-api-languageserver/src/test/java/org/eclipse/che/api/languageserver/ShowMessageRequestTest.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    public void testName() throws Exception {
+        CompletableFuture<MessageActionItem> future = new CompletableFuture<>();
+        when(transmitter.sendShowMessageRequest(any())).thenReturn(future);
+
+        ServerInitializerImpl initializer = new ServerInitializerImpl();
+        CheLanguageClient client = new CheLanguageClient(eventService, transmitter, "id");
+        CompletableFuture<Pair<LanguageServer, InitializeResult>> initialize = initializer.initialize(launcher, client, "/tmp");
+        Pair<LanguageServer, InitializeResult> resultPair = initialize.get();
+        server = resultPair.first;
+        Thread.sleep(1200);
```
Comment (skabashnyuk) [inline wsagent/che-core-api-languageserver/src/test/java/org/eclipse/che/api/languageserver/ShowMessageRequestTest.java]: what for?

**Comment by riuvshin** [comment]
maven profile that affects final assembly is not acceptable solution here due to many reasons, for example how you would test tagged version if that profile is not default and should be included only for testing purposes. We have to rework this somehow to be able to enable / disable this without rebuild.

**Review by dkulieshov — APPROVED** [review]

**Review by riuvshin — CHANGES_REQUESTED** [review]
we should not have that profile, as we discussed that should be reworked to be configurable.

**Review by vparfonov — APPROVED** [review]

**Review by riuvshin — APPROVED** [review]
please do not merge this before release 5.17.0 is done.  And also please create an issue to use this test LS in QA tests

**Review by tolusha — APPROVED** [review]

**Comment by riuvshin** [comment]
ci-build

**Pull request merged**